### PR TITLE
Flow: add otelcol.receiver.otlp component

### PR DESCRIPTION
**Pull request by rfratto**
Introduce an otelcol.receiver.otlp component which can receive OTLP data over gRPC or HTTP. There is more code here than would normally be necessary since I needed to introduce common types otelcol receiver components will share. 

This is a nearly 1:1 match to the upstream component with one exception: Custom authentication methods are not yet supported.

Future work will be needed to figure out how we want to expose authentication to components. The way it currently works in OpenTelemetry is by exposing an Authentication extension to components. We will likely need to do something different. Work on this will be done in a future PR.

Closes grafana/agent#2286.

### CODE REVIEW

**Comment by mattdurham** [comment]
I want to give this a closer look than I can at the moment but will give a deeper review tomorrow morning.

File: docs/sources/flow/reference/components/otelcol.receiver.otlp.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
+---
+aliases:
+- /docs/agent/latest/flow/reference/components/otelcol.receiver.otlp
+title: otelcol.receiver.otlp
+---
+
+# otelcol.receiver.otlp
```
Comment (rfratto) [inline docs/sources/flow/reference/components/otelcol.receiver.otlp.md:7]: I want to be able to give an example pipeline at the bottom of the page, but I can't at the moment, not until there's at least one exporter component.

**Review by tpaschalis — APPROVED** [review]
LGTM!

File: component/otelcol/config_grpc.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+type GRPCServerArguments struct {
+	Endpoint  string `river:"endpoint,attr,optional"`
+	Transport string `river:"transport,attr,optional"`
+
+	TLS *TLSServerArguments `river:"tls,block,optional"`
+
+	MaxRecvMsgSizeMiB    uint64 `river:"max_recv_msg_size_mib,attr,optional"`
+	MaxConcurrentStreams uint32 `river:"max_concurrent_streams,attr,optional"`
+	ReadBufferSize       int    `river:"read_buffer_size,attr,optional"`
+	WriteBufferSize      int    `river:"write_buffer_size,attr,optional"`
```
Comment (rfratto) [inline component/otelcol/config_grpc.go]: After a discussion offline, it'd be nice to have these be River strings which describe the unit (`"500B"`).

Ditto for other config types.

**Review by karengermond — APPROVED** [review]
LGTM!

**Pull request merged**